test/fail_compilation/retscope6.d

```diff
@@ -21,3 +21,120 @@ int* test() @safe
     arr[0] ~= &i;
     return arr[0][0];
 }
+
+/* TEST_OUTPUT:
+---
+fail_compilation/retscope6.d(7034): Error: reference to local variable `i` assigned to non-scope parameter `_param_1` calling retscope6.S.emplace!(int*).emplace
+fail_compilation/retscope6.d(7035): Error: reference to local variable `i` assigned to non-scope parameter `_param_0` calling retscope6.S.emplace2!(int*).emplace2
+fail_compilation/retscope6.d(7024): Error: scope variable `_param_2` assigned to `s` with longer lifetime
+fail_compilation/retscope6.d(7025): Error: scope variable `_param_2` assigned to `t` with longer lifetime
+fail_compilation/retscope6.d(7037): Error: template instance `retscope6.S.emplace4!(int*)` error instantiating
+---
+*/
+
+#line 7000
+
+alias T = int*;
+
+struct S
+{
+    T payload;
+
+    static void emplace(Args...)(ref S s, Args args) @safe
+    {
+        s.payload = args[0];
+    }
+
+    void emplace2(Args...)(Args args) @safe
+    {
+        payload = args[0];
+    }
+
+    static void emplace3(Args...)(S s, Args args) @safe
+    {
+        s.payload = args[0];
+    }
+
+    static void emplace4(Args...)(scope ref S s, scope out S t, scope Args args) @safe
```

Comment:
https://dlang.org/spec/attribute.html#scope

> scope cannot be applied to globals, statics, data members, ref or out parameters.

This is violating the spec.  The spec needs an update.

Reply:
https://dlang.org/spec/attribute.html#scope

> Assignment to a scope, other than initialization, is not allowed.

Also not accurate.

Reply:
As I've mentioned before, dip1000 has not been folded into the spec.

Reply:
Yes, and that's part of the problem.  Why hasn't it been done? And shouldn't it be done either prior to PRs like this or together with PRs like this?

Reply:
> Why hasn't it been done?

One large reason is because I can't get my spec fixes pulled.

Aside from that, DIP1000 despite being years old has little use, so I don't know how well it works. I can't get people to use DIP1000 features because phobos doesn't compile with dip1000. So I try to get phobos to compile with dip1000. I run into problems. I make corrections (like this PR). And I get blocked.

End result, no progress anywhere.
---
There is another issue. I used to work at Boeing in design. The designs ended up as drawings, using pen and ink on special plastic paper. Unsurprisingly, revising these drawings (including redrawing them) is quite expensive. But the designs changed constantly (bug fixes, improvements, changing requirements, the same issues that affect software). What was done was issuing DCN (Drawing Change Notices), which contained the fixes to the drawing. The official "drawing" was the original drawing plus a stack of DCNs, applied in chronological order (much like git, git is hardly a new concept). Eventually, the drawing may get redone with all the DCNs folded in.

The analogy for us is the spec, plus approved DIPs, plus Bugzilla Enhancement Requests and bug fixes. It's not ideal, but it's pragmatic. Halting all progress because it all hasn't been folded into the spec yet is not pragmatic, especially when my PRs to update the spec just sit there.

Reply:
> One large reason is because I can't get my spec fixes pulled.

I suggest putting a time limit on those PRs that haven't received feedback.  We already have a `72h no objection -> merge" label in the DMD repository, and I don't see why that can't be applied to your dlang.org PRs.  Noone has a right to complain if they chose not to participate.

> There is another issue. I used to work at Boeing in design...

I understand that, but I'm not going to be your rubber stamp.  I need to understand what these PRs are doing and why.

Others on this team are well within their charter to move this PR forward if they understand it and have confidence in the direction you're heading, but I'm not on board yet and have concerns.  I've already sent you an e-mail or two about the way `scope` is being inferred, and I don't like what I'm seeing in the implementation so far.  I think we need a way to set storage classes meta-programmatically, rather than reaching for inference.